OpenRA.Mods.Common/Traits/Buildings/DockClientManager.cs

```diff
@@ -0,0 +1,327 @@
+			dockClients = self.TraitsImplementing<IDockClient>().ToArray();
+		}
+
+		public Actor ReservedHostActor { get; protected set; }
+		public DockHost ReservedHost { get; protected set; }
+
+		DockHost lastReservedDockHost = null;
+		public DockHost LastReservedHost
+		{
+			get
+			{
+				if (lastReservedDockHost != null)
+				{
+					if (!lastReservedDockHost.IsEnabledAndInWorld)
+						lastReservedDockHost = null;
+					else
+						return lastReservedDockHost;
+				}
+
+				return ReservedHost;
+			}
+		}
+
+		public void UnreserveHost()
+		{
+			if (ReservedHost != null)
+			{
+				lastReservedDockHost = ReservedHost;
+				ReservedHost = null;
+				ReservedHostActor = null;
+				lastReservedDockHost.Unreserve(this);
+			}
+		}
+
+		/// <remarks> Additionaly returns true if reservation was succesful or we have already been reserved at this <see cref="DockHost"/>. </remarks>
+		public bool ReserveHost(Actor hostActor, DockHost host)
+		{
+			if (host == null)
+				return false;
+
+			if (ReservedHost == host)
+				return true;
+
+			UnreserveHost();
+			if (host.Reserve(hostActor, this))
+			{
+				ReservedHost = host;
+				ReservedHostActor = hostActor;
+
+				// After we have reserved a new Host we want to forget our old host.
+				lastReservedDockHost = null;
+				return true;
+			}
+
+			return false;
+		}
+
+		public void DockStarted(Actor self, Actor hostActor, DockHost host)
+		{
+			foreach (var client in dockClients)
+				client.DockStarted(self, hostActor, host);
+		}
+
+		public bool DockTick(Actor self, Actor hostActor, DockHost host)
+		{
+			if (IsTraitDisabled)
+				return true;
+
+			var cancel = true;
+			foreach (var client in dockClients)
+				if (!client.DockTick(self, hostActor, host))
+					cancel = false;
+
+			return cancel;
+		}
+
+		public void DockCompleted(Actor self, Actor hostActor, DockHost host)
+		{
+			foreach (var client in dockClients)
+				client.DockCompleted(self, hostActor, host);
+
+			UnreserveHost();
+		}
+
+		IEnumerable<IOrderTargeter> IIssueOrder.Orders
+		{
+			get
+			{
+				yield return new EnterAlliedActorTargeter<DockHostInfo>(
+					"ForceDock",
+					6,
+					Info.EnterCursor,
+					Info.EnterBlockedCursor,
+					DockingPossible,
+					target => CanDockAt(target, true, true));
+				yield return new EnterAlliedActorTargeter<DockHostInfo>(
+					"Dock",
+					5,
+					Info.EnterCursor,
+					Info.EnterBlockedCursor,
+					(actor, modifiers) => DockingPossible(actor),
+					target => CanDockAt(target, false, true));
+			}
+		}
+
+		void IResolveOrder.ResolveOrder(Actor self, Order order)
+		{
+			if (order.OrderString == "Dock")
+			{
+				var target = order.Target;
+
+				// Deliver orders are only valid for own/allied actors,
+				// which are guaranteed to never be frozen.
+				// TODO: support frozen actors
+				if (target.Type != TargetType.Actor)
+					return;
+
+				if (IsTraitDisabled)
+					return;
+
+				var dock = AvailableDockHosts(target.Actor, false, true).ClosestDock(self, this);
```

Comment:
One point here, there may be versions of docking where the physically closest dock is not the preferred one, for a variety of reasons (it could have a slower collection rate (different building type), might have a higher occupancy rate from other harvesters, etc.). Therefore to keep this generic, I think abstracting this to a `HighestPriorityDock() => ClosestDock()` method would be better, so that the highest priority can be defined by non-distance related factors if needed.

Reply:
Makes sense
It will need to be a feature anyways, you are supposed to be able to dock to allies, but prioritise your own docks

Reply:
I think it's better to implement with weights rather priorities

Reply:
I'd rather add this later on. It won't be used by harvesters and it's not a small feature

Reply:
That's fine, though I would add a "TO DO:" given the importance of the feature

Reply:
ok